### PR TITLE
CDAP-2764 use artifact repo in app deployment pipeline

### PR DESCRIPTION
**Pull request by albertshau**
Some refactoring to make the application deployment pipeline add
the jar to the artifact repo, then call the deployment pipeline
with the id for the newly added artifact. Unfortunately, the
old code path still exists because app templates rely on it, but
should be removed soon. ArtifactID is now also part of the app
spec, which is a requirement for plugin integration with apps.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT2448

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 1]
```diff
@@ -258,6 +272,16 @@ private BodyConsumer deployApplication(final HttpResponder responder,
       return null;
     }
 
+    // error handling needs to be refactored here, should be able just to throw the exception, but the caller
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: may be prefix it with TODO ?

File: cdap-client-tests/src/test/java/co/cask/cdap/client/PreferencesClientTestRun.java [hunk 2]
```diff
@@ -238,7 +239,11 @@ public void testPreferences() throws Exception {
       Assert.assertEquals(propMap, client.getApplicationPreferences(FAKE_APP_ID, false));
       Assert.assertEquals(propMap, client.getProgramPreferences(flow, false));
     } finally {
-      appClient.delete(FAKE_APP_ID);
+      try {
+        appClient.delete(FAKE_APP_ID);
+      } catch (ApplicationNotFoundException e) {
+        // ok if this happens, means its already deleted.
```
Comment (anwar6953) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/PreferencesClientTestRun.java]: why would it already be deleted, if its redeployed line 237?

Comment (albertshau) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/PreferencesClientTestRun.java]: its in a finally, so there could be an exception in the test that happens before its redeployed. This test will fail, but we'd want the failure exception to be from that earlier failure and not from this delete.

Comment (anwar6953) [inline cdap-client-tests/src/test/java/co/cask/cdap/client/PreferencesClientTestRun.java]: I see. ok.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java [hunk 3]
```diff
@@ -85,6 +87,12 @@ public DefaultAppConfigurer(Application app, String configuration) {
     this.configuration = configuration;
   }
 
+  public DefaultAppConfigurer(Id.Artifact artifactId, Application app, String configuration) {
+    this(app);
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java]: this(app, configuration);

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 4]
```diff
@@ -371,6 +412,30 @@ private void getAppDetails(HttpResponder responder, String namespace, String nam
     }
   }
 
+  // parse a filename into an Id.Artifact, assuming the filename is of format {name}-{version}.jar
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: Will this be the only way to get the version of the jar? Will we also have a fall back option of trying to get it from the manifest?

Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: yeah i will need to add that logic, but we should try and get from manifest if the header is not present. Opened CDAP-3310

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/LocalArtifactLoaderStage.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.twill.filesystem.Location;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.Nullable;
+
+/**
+ * LocalArchiveLoaderStage gets a {@link Location} and emits a {@link ApplicationDeployable}.
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/LocalArtifactLoaderStage.java]: LocalArtifactLoaderStage

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java [hunk 6]
```diff
@@ -186,12 +186,40 @@ public ArtifactDetail getArtifact(Id.Artifact artifactId) throws IOException, Ar
...
    * @param artifactId the id of the artifact to inspect and store
    * @param artifactFile the artifact to inspect and store
+   * @return detail about the newly added artifact
+   * @throws IOException if there was an exception reading from the artifact store
+   * @throws WriteConflictException if there was a write conflict writing to the ArtifactStore
+   * @throws ArtifactAlreadyExistsException if the artifact already exists
+   * @throws InvalidArtifactException if the artifact is invalid. For example, if it is not a zip file,
+   *                                  or the application class given is not an Application.
+   */
+  public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile)
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java]: can't we just call addArtifact(artifactId, artifactFile, null) ?

**Comment by gokulavasan** [comment]
Few comments. LGTM otherwise

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java (hunk 6 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java]: functionally yes, but this method doesn't throw an exception about parent artifacts not found whereas that one does.

Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java]: Oh ok.

**Comment by gokulavasan** [comment]
LGTM :+1:

**Pull request merged**